### DISCUSSION TITLE
How to embed IL code in .net5 ?

### DISCUSSION THREAD

**Discussion by Nepcix** [dotnet/roslyn#60792] [Q&A]
I have a problem and I got some solutions from this https://stackoverflow.com/questions/17270429/forcing-hardware-accelerated-rendering
one of the answers need IL code，but it only works in .net framework
is there any way to embed IL code into .net5 assembly ?
or
the key of this problem is how to export variable NvOptimusEnablement and AmdPowerXpressRequestHighPerformance 
can c# export variable like c++ ?

**Comment by PathogenDavid** [comment]
TL;DR: That solution does not work for modern .NET and there is no simple solution if you care about AMD PowerXpress.

The way the `.exe` works between .NET Framework and modern .NET is fundamentally different. In .NET Framework the `.exe` actually contained IL code, on .NET Core and later it's a tiny AppHost executable that bootstraps the runtime and loads you application. (You can find the [source of the AppHost in the runtime repo](https://github.com/dotnet/runtime/blob/d91e8217d5579f29b2fc085a6adf7d09cba1d9b5/src/native/corehost/corehost.cpp).)

As such, even if you injected that IL into your app after building it's not going to do what you want.

----------

For the Nvidia Optimus hint, you can just dynamically load `nvapi` before initializing any graphics stuff and it will work as expected:

```csharp
NativeLibrary.TryLoad(Environment.Is64BitProcess ? "nvapi64.dll" : "nvapi.dll", out _);
```

([The documentation](https://docs.nvidia.com/gameworks/content/technologies/desktop/optimus.htm) implies this has to be a static link, but I've tested this on an Optimus device and it works as expected.)

---------

Unfortunately I'm not aware of a similar solution for AMD PowerXpress. I've written in my notes to investigate using [AMD Display Library](https://gpuopen.com/adl/) to configure [the settings mentioned here](https://gpuopen.com/learn/amdpowerxpressrequesthighperformance/), but I don't have access to an AMD PowerXpress device to test with so I've put off doing it.

(Note that AMD says that the PowerXpress hint isn't necessary if you're using Vulkan.)

---------

It's also worth noting that if your app is DirectX-based, you actually don't even need these hints anymore if you're targeting Windows 10 1803 or later because you can just use [`IDXGIFactory6::EnumAdapterByGpuPreference method`](https://docs.microsoft.com/en-us/windows/win32/api/dxgi1_6/nf-dxgi1_6-idxgifactory6-enumadapterbygpupreference).

-------------

If you want to insist on using the same performance hints as C++, you could make a custom AppHost with these hints exported. Unfortunately I don't know of an easy-to-follow guide on how to do this, just that it's possible. (You might find [my notes on how the default AppHost gets generated](https://github.com/dotnet/roslyn/discussions/48273#discussioncomment-115815) to be helpful if you go down this route.)
